File: CPMS12m/Lab7/Dictionary.java
//-----------------------------------------------------------------------------
//Zachary Levenberg
//ID: 1049451
//12b
//5/27/15
//This file implements a lookup dictionary using a binary search tree
// Dictionary.java
//-----------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface{

	private class Node{
		String key;
		String value;
		Node left;
		Node right;

		Node(String k, String v){
			key = k;
			value = v;
			left = null;
			right = null;
		}
	}

	private Node root;
	private int numItems;

	public Dictionary(){
		root = null;
		numItems = 0;
	}

	//private functions-------------------------------------------------------------

	private Node findKey(Node R, String k){
		if((R==null) || k.equals(R.key)){
			return R;
		}
		if(k.compareTo(R.key) < 0){
			return findKey(R.left, k);
		} else{
			return findKey(R.right, k);
		}
	}

	// findLeftmost()
	// returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
	private Node findLeftmost(Node R){
		Node L = R;
		if( L!=null ) for( ; L.left!=null; L=L.left) ;
		return L;
	}


	// findParent()
	// returns the parent of N in the subtree rooted at R, or returns NULL 
	// if N is equal to R. (pre: R != NULL)
	private Node findParent(Node N, Node R){
		Node P = null;
		if(N!=R){
			P = R;
			while(P.left!=N && P.right!= N){
				if(N.key.compareTo(P.key) < 0){
					P = P.left;
				} else {
					P = P.right;
				}
			}
		}
		return P;
	}


	// printInOrder()
	// prints the (key, value) pairs belonging to the subtree rooted at R in order
	// of increasing keys to file pointed to by out.
	private void printInOrder(StringBuffer sb, Node R){
		if( R!=null ){
			printInOrder(sb, R.left);
			sb.append(R.key).append(" ").append(R.value).append("\n");
			printInOrder(sb, R.right);	   
		}
	}

	//public functions-----------------------------------------------------------

	// isEmpty()
	// returns true if this Dictionary is empty, false otherwise
	// pre: none
	public boolean isEmpty(){
		if(root==null){
			return true;
		} else {
			return false;
		}
	}

	// size()
	// returns the number of entries in this Dictionary
	// pre: none
	public int size(){
		return numItems;
	}

	// lookup()
	// returns value associated key, or null reference if no such key exists
	// pre: none
	public String lookup(String key){
		Node c;
		c = findKey(root, key);
		if(c==null){
			return null;
		} else{
			return c.value;
		}
	}

	// insert()
	// inserts new (key,value) pair into this Dictionary
	// pre: lookup(key)==null
	public void insert(String key, String value) throws KeyCollisionException{
		Node N, A, B;
		if( findKey(root, key)!=null ){
			throw new KeyCollisionException("cannot insert duplicate keys");		
		}
		N = new Node(key, value);
		B = null;
		A = root;
		while( A!=null ){
			B = A;
			if( key.compareTo(A.key)<0 ) A = A.left;
			else A = A.right;
		}
		if( B==null ) root = N;
		else if( key.compareTo(B.key)<0) B.left = N;
		else B.right = N;
		numItems++;
	}

	// delete()
	// deletes pair with the given key
	// pre: lookup(key)!=null
	public void delete(String key) throws KeyNotFoundException{
		Node N, P, S;
		N = findKey(root, key);
		if( N==null ){
			throw new KeyNotFoundException("cannot delete non-existent found");
		}
		if( N.left==null && N.right==null ){
			if( N==root ){
				root = null;
			}else{
				P = findParent(N, root);
				if( P.right==N ) P.right = null;
				else P.left = null;
			}
		}else if( N.right==null ){
			if( N==root ){
				root = N.left;
			}else{
				P = findParent(N, root);
				if( P.right==N ) P.right = N.left;
				else P.left = N.left;
			}
		}else if( N.left==null ){
			if( N==root ){
				root = N.right;
			}else{
				P = findParent(N, root);
				if( P.right==N ) P.right = N.right;
				else P.left = N.right;
			}
		}else{  // N->left!=NULL && N->right!=NULL
			S = findLeftmost(N.right);
			N.key = S.key;
			N.value = S.value;
			P = findParent(S, N);
			if( P.right==S ) P.right = S.right;
			else P.left = S.right;
		}
		numItems--;
	}

	// makeEmpty()
	// pre: none
	public void makeEmpty(){
		root = null;
		numItems = 0;
	}

	// toString()
	// returns a String representation of this Dictionary
	// overrides Object's toString() method
	// pre: none
	public String toString(){
		StringBuffer sb = new StringBuffer();
		printInOrder(sb, root);
		return new String(sb);
	}

}

File: CPMS12m/Lab7/Dictionary.c
//-----------------------------------------------------------------------------
// Dictionary.c
// Binary Search Tree implementation of the Dictionary ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

// private types and functions ------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value;
   struct NodeObj* left;
   struct NodeObj* right;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor for private Node type
Node newNode(char* k, char* v) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->key = k;
   N->value = v;
   N->left = N->right = NULL;
   return(N);
}

// freeNode()
// destructor for private Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// DictionaryObj
typedef struct DictionaryObj{
   Node root;
   int numPairs;
} DictionaryObj;


// findKey()
// returns the Node containing key k in the subtree rooted at R, or returns
// NULL if no such Node exists
Node findKey(Node R, char* k){
   if(R==NULL || strcmp(k, R->key)==0) 
      return R;
   if( strcmp(k, R->key)<0 ) 
      return findKey(R->left, k);
   else // strcmp(k, R->key)>0
      return findKey(R->right, k);
}


// findParent()
// returns the parent of N in the subtree rooted at R, or returns NULL 
// if N is equal to R. (pre: R != NULL)
Node findParent(Node N, Node R){
   Node P=NULL;
   if( N!=R ){
      P = R;
      while( P->left!=N && P->right!=N ){
         if(strcmp(N->key, P->key)<0)
            P = P->left;
         elsehttps://classes.soe.ucsc.edu/cmps012b/Spring15/Examples/Labs/lab7/Makefile
            P = P->right;
      }
   }
   return P;
}

// findLeftmost()
// returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
Node findLeftmost(Node R){
   Node L = R;
   if( L!=NULL ) for( ; L->left!=NULL; L=L->left) ;
   return L;
}

// printInOrder()
// prints the (key, value) pairs belonging to the subtree rooted at R in order
// of increasing keys to file pointed to by out.
void printInOrder(FILE* out, Node R){
   if( R!=NULL ){
      printInOrder(out, R->left);
      fprintf(out, "%s %s\n", R->key, R->value);
      printInOrder(out, R->right);
   }
}

// deleteAll()
// deletes all Nodes in the subtree rooted at N.
void deleteAll(Node N){
   if( N!=NULL ){
      deleteAll(N->left);
      deleteAll(N->right);
      freeNode(&N);
   }
}

// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->root = NULL;
   D->numPairs = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){https://classes.soe.ucsc.edu/cmps012b/Spring15/Examples/Labs/lab7/Makefile
      makeEmpty(*pD);
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numPairs==0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling size() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numPairs);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
   Node N;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   N = findKey(D->root, k);
   return ( N==NULL ? NULL : N->value );
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
   Node N, A, B;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling insert() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( findKey(D->root, k)!=NULL ){
      fprintf(stderr, 
         "Dictionary Error: cannot insert() duplicate key: \"%s\"\n", k);
      exit(EXIT_FAILURE);
   }
   N = newNode(k, v);
   B = NULL;
   A = D->root;
   while( A!=NULL ){
      B = A;
      if( strcmp(k, A->key)<0 ) A = A->left;
      else A = A->right;
   }
   if( B==NULL ) D->root = N;
   else if( strcmp(k, B->key)<0 ) B->left = N;
   else B->right = N;
   D->numPairs++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
   Node N, P, S;
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   N = findKey(D->root, k);
   if( N==NULL ){
      fprintf(stderr, 
         "Dictionary Error: cannot delete() non-existent key: \"%s\"\n", k);
      exit(EXIT_FAILURE);
   }
   if( N->left==NULL && N->right==NULL ){
      if( N==D->root ){
         D->root = NULL;
         freeNode(&N);
      }else{
         P = findParent(N, D->root);
         if( P->right==N ) P->right = NULL;
         else P->left = NULL;
         freeNode(&N);
      }
   }else if( N->right==NULL ){
      if( N==D->root ){
         D->root = N->left;
         freeNode(&N);
      }else{
         P = findParent(N, D->root);
         if( P->right==N ) P->right = N->left;
         else P->left = N->left;
         freeNode(&N);
      }
   }else if( N->left==NULL ){
      if( N==D->root ){
         D->root = N->right;
         freeNode(&N);
      }else{
         P = findParent(N, D->root);
         if( P->right==N ) P->right = N->right;
         else P->left = N->right;
         freeNode(&N);
      }
   }else{  // N->left!=NULL && N->right!=NULL
      S = findLeftmost(N->right);
      N->key = S->key;
      N->value = S->value;
      P = findParent(S, N);
      if( P->right==S ) P->right = S->right;
      else P->left = S->right;
      freeNode(&S);
   }
   D->numPairs--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
   deleteAll(D->root);
   D->root = NULL;
   D->numPairs = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling printDictionary() on NULL"
         " Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   printInOrder(out, D->root);
}

